### PR TITLE
Fix custom RPC node entry

### PR DESCRIPTION
**Pull request by JamesSmartCell**
Fix vulnerability where malicious RPC could be entered via a website inject.

### CODE REVIEW

File: app/src/main/java/com/alphawallet/app/ui/DappBrowserFragment.java [hunk 1]
```diff
@@ -1438,6 +1447,24 @@ private void txError(Throwable throwable)
             confirmationDialog.dismiss();
     }
 
+    //Display error
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/ui/DappBrowserFragment.java:1450]: I think this comment is redundant as the method name is clear enough.

File: app/src/main/java/com/alphawallet/app/viewmodel/DappBrowserViewModel.java [hunk 2]
```diff
@@ -350,26 +351,63 @@ public String getSessionId(String url)
         return Uri.parse(uriString).getUserInfo();
     }
 
-    public void addCustomChain(WalletAddEthereumChainObject chainObject) {
-        this.ethereumNetworkRepository.saveCustomRPCNetwork(chainObject.chainName, extractRpc(chainObject), chainObject.getChainId(),
-                chainObject.nativeCurrency.symbol, "", "", false, -1L);
+    public boolean addCustomChain(WalletAddEthereumChainObject chainObject) {
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/DappBrowserViewModel.java:354]: Looks like not formatted as the `{` should be on the next line.

File: app/src/main/java/com/alphawallet/app/viewmodel/DappBrowserViewModel.java [hunk 3]
```diff
@@ -350,26 +351,63 @@ public String getSessionId(String url)
...
                 .isDisposed();
+
+        return true;
+    }
+
+    private String extractBlockExplorer(WalletAddEthereumChainObject chainObject)
+    {
+        for (String thisRpc : chainObject.blockExplorerUrls)
+        {
+            if (isValidUrl(thisRpc)) //ensure RPC doesn't contain malicious code
```
Comment (seabornlee) [inline app/src/main/java/com/alphawallet/app/viewmodel/DappBrowserViewModel.java:375]: Suggest adding unit test for `isValidUrl` with the malicious payload to ensure it won't pass the gate.

**Review by seabornlee — CHANGES_REQUESTED** [review]

**Comment by JamesSmartCell** [comment]
Thanks @seabornlee all good observations!

**Review by seabornlee — APPROVED** [review]

**Pull request merged**